mlflow/tracking/client.py

```diff
@@ -1536,16 +1555,45 @@ def log_image(
                 "`key` to log dynamic image charts or `artifact_file` for saving static images. "
             )
 
+        import numpy as np
+
+        # Convert image type to PIL if its a numpy array
+        if isinstance(image, np.ndarray):
+            image = _convert_numpy_to_pil_image(image)
+        elif isinstance(image, Image):
+            image = image.to_pil()
+
         if artifact_file is not None:
-            self._log_image_as_artifact(run_id, image, artifact_file)
+            with self._log_artifact_helper(run_id, artifact_file) as tmp_path:
+                image.save(tmp_path)
 
         elif key is not None:
+            # Check image key for invalid characters
+            if not re.match(r"^[a-zA-Z0-9_\-./ ]+$", key):
+                raise ValueError(
+                    "The `key` parameter may only contain alphanumerics, underscores (_), "
+                    "dashes (-), periods (.), spaces ( ), and slashes (/)."
+                )
+
             step = step or 0
             timestamp = timestamp or get_current_time_millis()
-            filename = f"images/{key}/{key}_step_{step}_{uuid.uuid4()}"
+
+            # Sanitize key to use in filename (replace / with # to avoid subdirectories)
+            sanitized_key = re.sub(r"/", "#", key)
+            filename = f"images/{sanitized_key}%step%{step}%timestamp%{timestamp}%{uuid.uuid4()}"
+
+            # Save full-res image
             image_filepath = f"{filename}.png"
+            with self._log_artifact_helper(run_id, image_filepath) as tmp_path:
```

Comment:
curious - why are we saving both the original and compressed image?

Reply:
ping this one.

Reply:
Sorry, missed this 😅 original image would be the full-resolution one when you full-screen an image, and the compressed image is saved so we can load it quickly in the slider.